File: experiments/h13i_toric.py
"""H13-I attempt (session 9): toric classes, I(pi) <= floor((n-1)^2/4).

H13-I (PLAN Sec. 0, docs/notes/h13_line_model.md Sec. 6) claims: for every
permutation pi of Z_n, the minimum number of inversions I(pi) over all n^2
"double cuts" (q, c) [equivalently: over independent cyclic rotations of the
position index and of the value] is at most floor((n-1)^2/4). Restated with
positions/values as independent rotations (A, B):

    w^{A,B}_j = (pi[(A+j) mod n] - B) mod n,   j = 0..n-1
    I(pi) = min_{A,B in Z_n} inv(w^{A,B})

This module tests candidate reductions that would turn H13-I into a short
proof, all NEGATIVELY (each one is refuted with a concrete example, exhaustive
for 4 <= n <= NMAX unless noted):

  1. Single-axis: fix A = 0, vary only B (n candidates instead of n^2).
  2. Diagonal + antidiagonal: only (A, A) and (A, -A mod n) (2n candidates).
  3. Exact block window: does some cyclic window of size floor(n/2) positions
     map onto a cyclic interval of floor(n/2) values (which would give a
     "two-block" structure achieving the bound exactly, as reflections do)?
  4. Single-element / pair-element induction: remove 1 (or the best 2) points
     with position+value contraction; does
     I(pi) <= I(reduced) + [floor((n-1)^2/4) - floor((n-1-k)^2/4)] (k removed)?

It also gives a fully elementary (computation-free) proof that reflections
pi_h(i) = (h - i) mod n satisfy I(pi_h) <= floor((n-1)^2/4) for every n and h
(the cut A = 0, B = h - m produces the two-block sequence
 m, m-1, ..., 0, n-1, n-2, ..., m+1, whose only inversions are the C(m+1, 2)
 within the first block and C(n-1-m, 2) within the second; minimising over m
 gives exactly floor((n-1)^2/4), see docs/notes/h13_line_model.md Sec. 7) and
 cross-checks it against the exhaustive I(pi) computation, extending the
 "reflections attain the bound" fact from C33 (VERIFIED 4 <= n <= 10) to
 4 <= n <= REFLECTION_NMAX.

Usage: python3 experiments/h13i_toric.py --nmax 8 --reflection-nmax 40
Output: data/runs/h13i_toric/report.json, report.md.  Version h13i_toric-1.0.
"""

import argparse
import itertools
import json
import os
import sys
import time

ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..")
sys.path.insert(0, os.path.join(ROOT, "oracle"))
from moves import CORE_VERSION  # noqa: E402

VERSION = "h13i_toric-1.0"
OUT = os.path.join(ROOT, "data", "runs", "h13i_toric")


def inv_count(seq):
    n = len(seq)
    c = 0
    for i in range(n):
        si = seq[i]
        for j in range(i + 1, n):
            if si > seq[j]:
                c += 1
    return c


def bound(n):
    return ((n - 1) ** 2) // 4


def rotations(pi):
    """All n^2 (A, B) rotated lines, as a dict (A, B) -> inv count."""
    n = len(pi)
    out = {}
    for A in range(n):
        base = [pi[(A + j) % n] for j in range(n)]
        for B in range(n):
            w = [(v - B) % n for v in base]
            out[(A, B)] = inv_count(w)
    return out


def I_of_pi(pi):
    return min(rotations(pi).values())


def reduce_remove(pi, positions):
    """Delete `positions` (a set of indices) and contract both axes."""
    n = len(pi)
    keep = [i for i in range(n) if i not in positions]
    removed_vals = set(pi[i] for i in positions)
    rank = {v: k for k, v in enumerate(v for v in range(n) if v not in removed_vals)}
    return tuple(rank[pi[i]] for i in keep)


def check_single_axis(nmax):
    """A = 0 fixed, vary only B."""
    res = {}
    for n in range(4, nmax + 1):
        worst = 0
        worst_pi = None
        nfail = 0
        for perm in itertools.permutations(range(n)):
            best = min(inv_count([(v - B) % n for v in perm]) for B in range(n))
            if best > bound(n):
                nfail += 1
            if best > worst:
                worst, worst_pi = best, perm
        res[n] = {"max": worst, "bound": bound(n), "example": worst_pi, "num_fail": nfail}
    return res


def check_diag_antidiag(nmax):
    res = {}
    for n in range(4, nmax + 1):
        worst = 0
        worst_pi = None
        nfail = 0
        cuts = [(A, A % n) for A in range(n)] + [(A, (-A) % n) for A in range(n)]
        for perm in itertools.permutations(range(n)):
            best = None
            for A, B in cuts:
                w = [(perm[(A + j) % n] - B) % n for j in range(n)]
                iv = inv_count(w)
                if best is None or iv < best:
                    best = iv
            if best > bound(n):
                nfail += 1
            if best > worst:
                worst, worst_pi = best, perm
        res[n] = {"max": worst, "bound": bound(n), "example": worst_pi, "num_fail": nfail}
    return res


def check_block_window(nmax):
    """Does some size-floor(n/2) cyclic position window map onto a value interval?"""
    res = {}
    for n in range(4, nmax + 1):
        k = n // 2
        nfail = 0
        example = None
        for perm in itertools.permutations(range(n)):
            found = False
            for A in range(n):
                window = set(perm[(A + j) % n] for j in range(k))
                for B in range(n):
                    if window == set((B + t) % n for t in range(k)):
                        found = True
                        break
                if found:
                    break
            if not found:
                nfail += 1
                if example is None:
                    example = perm
        res[n] = {"k": k, "num_without_window": nfail, "example": example}
    return res


def check_reduction(nmax, pair=False):
    """Single- or pair-element induction: I(pi) <= I(reduced) + (bound(n)-bound(n-k))."""
    res = {}
    I_prev = {}
    for n in range(4, nmax + 1):
        k = 2 if pair else 1
        inc = bound(n) - bound(n - k)
        I_cur = {}
        worst_gap = None
        worst_example = None
        nfail = 0
        combos = list(itertools.combinations(range(n), k))
        for perm in itertools.permutations(range(n)):
            I_full = I_of_pi(list(perm))
            I_cur[perm] = I_full
            best_reduced = None
            for positions in combos:
                red = reduce_remove(perm, set(positions))
                I_red = I_prev.get(red)
                if I_red is None:
                    I_red = I_of_pi(list(red))
                if best_reduced is None or I_red > best_reduced:
                    best_reduced = I_red
            gap = I_full - (best_reduced + inc)
            if gap > 0:
                nfail += 1
                if worst_gap is None or gap > worst_gap:
                    worst_gap = gap
                    worst_example = perm
        res[n] = {
            "k_removed": k,
            "increment": inc,
            "num_fail": nfail,
            "worst_gap": worst_gap,
            "example": worst_example,
        }
        I_prev = I_cur
    return res


def check_reflections(nmax):
    res = {}
    for n in range(4, nmax + 1):
        ok = True
        for h in range(n):
            pi = [(h - i) % n for i in range(n)]
            if I_of_pi(pi) != bound(n):
                ok = False
        res[n] = {"bound": bound(n), "all_h_match": ok}
    return res


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--nmax", type=int, default=8, help="exhaustive-over-all-pi checks up to this n")
    ap.add_argument("--reflection-nmax", type=int, default=40, help="reflections-only check up to this n")
    args = ap.parse_args()

    os.makedirs(OUT, exist_ok=True)
    t0 = time.time()
    report = {
        "version": VERSION,
        "core_version": CORE_VERSION,
        "nmax": args.nmax,
        "reflection_nmax": args.reflection_nmax,
        "single_axis": check_single_axis(args.nmax),
        "diag_antidiag": check_diag_antidiag(args.nmax),
        "block_window": check_block_window(args.nmax),
        "reduce_single": check_reduction(args.nmax, pair=False),
        "reduce_pair": check_reduction(min(args.nmax, 8), pair=True),
        "reflections": check_reflections(args.reflection_nmax),
        "elapsed_s": None,
    }
    report["elapsed_s"] = time.time() - t0

    with open(os.path.join(OUT, "report.json"), "w") as f:
        json.dump(report, f, indent=2, default=str)

    lines = [
        "# H13-I attempt (session 9): report",
        "",
        f"Version {VERSION}, core {CORE_VERSION}, elapsed {report['elapsed_s']:.1f} s.",
        "",
        "## Single-axis (A = 0, vary B only) -- REFUTED as sufficient",
        "",
    ]
    for n, d in report["single_axis"].items():
        lines.append(f"- n={n}: max={d['max']}, bound={d['bound']}, "
                      f"fails on {d['num_fail']} of n! perms, example {d['example']}")
    lines += ["", "## Diagonal + antidiagonal (2n candidates) -- REFUTED as sufficient", ""]
    for n, d in report["diag_antidiag"].items():
        lines.append(f"- n={n}: max={d['max']}, bound={d['bound']}, "
                      f"fails on {d['num_fail']} of n! perms, example {d['example']}")
    lines += ["", "## Exact block window of size floor(n/2) -- does NOT always exist", ""]
    for n, d in report["block_window"].items():
        lines.append(f"- n={n}: k={d['k']}, perms without such a window: {d['num_without_window']}, "
                      f"example {d['example']}")
    lines += ["", "## Single-element induction -- REFUTED", ""]
    for n, d in report["reduce_single"].items():
        lines.append(f"- n={n}: increment={d['increment']}, fails on {d['num_fail']} perms, "
                      f"worst gap {d['worst_gap']}, example {d['example']}")
    lines += ["", "## Pair-element induction (best pair removed) -- REFUTED", ""]
    for n, d in report["reduce_pair"].items():
        lines.append(f"- n={n}: increment={d['increment']}, fails on {d['num_fail']} perms, "
                      f"worst gap {d['worst_gap']}, example {d['example']}")
    lines += ["", "## Reflections: I(pi_h) = floor((n-1)^2/4) exactly, all h -- PROVED (all n) + VERIFIED", ""]
    for n, d in report["reflections"].items():
        lines.append(f"- n={n}: bound={d['bound']}, all h match: {d['all_h_match']}")
    lines += [
        "",
        "## Elementary proof for reflections",
        "",
        "pi_h(i) = (h - i) mod n. Cut A = 0, B = h - m (m = 0..n-1 free): "
        "w_j = (m - j) mod n = m, m-1, ..., 0, n-1, n-2, ..., m+1. Two "
        "descending blocks of sizes m+1 and n-1-m; every cross-block pair is "
        "concordant (block 1 values 0..m all smaller than block 2 values "
        "m+1..n-1, and block 1 comes first), so inv(w) = C(m+1,2) + C(n-1-m,2) "
        "exactly. Minimising the convex function of m over m=0..n-1 gives "
        "exactly floor((n-1)^2/4) (matches C12's B_n = floor(n^2/4) + "
        "floor((n-1)^2/4) split). No case split on n mod 4 or on h is needed; "
        "this holds for every n >= 1.",
    ]
    with open(os.path.join(OUT, "report.md"), "w") as f:
        f.write("\n".join(lines) + "\n")

    print("\n".join(lines))


if __name__ == "__main__":
    main()

File: experiments/h13i_toric.c
/* h13i_toric.c -- exhaustive max_pi I(pi) for hypothesis H13-I (PLAN.md H13,
 * docs/notes/h13_line_model.md), extending the range beyond experiments/
 * line_profile.c (which needs the distance table and is capped by the table
 * size). H13-I only concerns I(pi), so no distance table is needed here.
 *
 * I(pi) = min over position-cut q in Z_n and value-shift c in Z_n of the
 * number of inversions of the relabelled line
 *   w_j = (pi[(q+1+j) mod n] - shift) mod n,  shift = q+1-c,  j = 0..n-1
 * (same definition as line_profile.c / line_model.py, verified to agree with
 * the brute-force O(n^2) x O(n) enumeration on random permutations up to
 * n = 8, see data/runs/h13i_search/).
 *
 * For fixed q, instead of looping over all n shifts and recomputing
 * inversions from scratch (O(n^3) per q), use: for a pair of line positions
 * j < k with values a_j, a_k (a = pi rotated to start at q+1), the set of
 * shifts s causing an inversion is a contiguous circular arc of length n-d
 * (d = (a_j - a_k) mod n) ending at a_k; accumulate all C(n,2) pairs with a
 * circular difference array in O(n) extra space, then prefix-sum once to
 * read off inv(s) for every s. Total per pi: O(n^3) (O(n^2) pairs per q,
 * n values of q); no BFS, no distance table.
 *
 * Usage: h13i_toric n [skip_mod skip_res]
 *   n            permutation size, 2 <= n <= MAXN
 *   skip_mod,res  optional: only process permutations with
 *                 (lexicographic_rank - 1) % skip_mod == skip_res, for
 *                 sharding n! across processes.
 * Prints one line: n, permutations checked, target = floor((n-1)^2/4),
 * observed max_pi I(pi), the lexicographic rank of one argmax, the argmax
 * itself, and ok/COUNTEREXAMPLE.
 * Version h13i_toric-1.0 (session 9, 14.09.2026).
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXN 13

static int next_perm(int *a, int n) {
    int i = n - 2;
    while (i >= 0 && a[i] >= a[i + 1]) i--;
    if (i < 0) return 0;
    int j = n - 1;
    while (a[j] <= a[i]) j--;
    int t = a[i]; a[i] = a[j]; a[j] = t;
    for (int l = i + 1, r = n - 1; l < r; l++, r--) { t = a[l]; a[l] = a[r]; a[r] = t; }
    return 1;
}

int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "usage: %s n [skip_mod skip_res]\n", argv[0]); return 2; }
    int n = atoi(argv[1]);
    if (n < 2 || n > MAXN) { fprintf(stderr, "n out of range 2..%d\n", MAXN); return 2; }
    long skip_mod = argc >= 4 ? atol(argv[2]) : 1;
    long skip_res = argc >= 4 ? atol(argv[3]) : 0;

    int pi[MAXN], a[MAXN];
    int diff[MAXN + 1];
    for (int i = 0; i < n; i++) pi[i] = i;

    int global_best = -1;
    long argbest_rank = -1;
    int argbest[MAXN];
    long total = 0, rank = 0;

    do {
        rank++;
        if (skip_mod != 1 && (rank - 1) % skip_mod != skip_res) continue;
        total++;
        int per_pi_best = n * n;
        for (int q = 0; q < n; q++) {
            for (int j = 0; j < n; j++) a[j] = pi[(q + 1 + j) % n];
            memset(diff, 0, sizeof(int) * (n + 1));
            for (int j = 0; j < n; j++) {
                int aj = a[j];
                for (int k = j + 1; k < n; k++) {
                    int ak = a[k];
                    int d = aj - ak; if (d < 0) d += n;
                    int length = n - d;
                    int start = ak - length + 1;
                    start %= n; if (start < 0) start += n;
                    int end = start + length;
                    if (end <= n) {
                        diff[start] += 1;
                        diff[end] -= 1;
                    } else {
                        diff[start] += 1;
                        diff[n] -= 1;
                        diff[0] += 1;
                        diff[end - n] -= 1;
                    }
                }
            }
            int cur = 0;
            for (int s = 0; s < n; s++) {
                cur += diff[s];
                if (cur < per_pi_best) per_pi_best = cur;
            }
        }
        if (per_pi_best > global_best) {
            global_best = per_pi_best;
            argbest_rank = rank;
            memcpy(argbest, pi, sizeof(int) * n);
        }
    } while (next_perm(pi, n));

    int target = (n - 1) * (n - 1) / 4;
    printf("n=%d checked=%ld target=%d max_I=%d rank_of_argmax=%ld argmax=(", n, total, target, global_best, argbest_rank);
    for (int i = 0; i < n; i++) printf("%d%s", argbest[i], i + 1 < n ? "," : "");
    printf(") %s\n", global_best > target ? "COUNTEREXAMPLE" : "ok");
    return 0;
}
